### PR TITLE
storage: shuffle replicas for scanner iteration

### PR DESCRIPTION
**Pull request by petermattis**
TestRebalance_3To5Small was often failing to converge because the
replica scanners on the initial 3 nodes were scanning over replicas in
lock step. If one node started scanning slightly before the others it
would encounter the range first and acquire the leader lease which would
prohibit the replica from being rebalanced away from that node.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9703)

### CODE REVIEW

File: storage/store.go [hunk 1]
```diff
@@ -224,6 +226,14 @@ func newStoreRangeSet(store *Store) *storeRangeSet {
 	}
 }
 
+func newStoreShuffledRangeSet(store *Store) *storeRangeSet {
```
Comment (tschottdorf) [inline storage/store.go]: `newShuffledStoreRangeSet`?

File: storage/store.go [hunk 2]
```diff
@@ -224,6 +226,14 @@ func newStoreRangeSet(store *Store) *storeRangeSet {
...
+	return &storeRangeSet{
+		store:   store,
+		visited: -1,
+		shuffle: true,
+	}
```
Comment (tschottdorf) [inline storage/store.go]: For future-proofing, better call `newStoreRangeSet` and then set `shuffled: true`.

**Review by tbg — APPROVED** [review]
LGTM

File: storage/store.go (hunk 2 above)
Comment (petermattis) [inline storage/store.go]: Done.

File: storage/store.go (hunk 1 above)
Comment (petermattis) [inline storage/store.go]: Sure.

**Comment by cuongdo** [comment]
:lgtm: 

I assume you've run the 3To5Small test on this code?

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9703#-:-KTEiwe8FdCMzZCoIFhJ:bgwdfj6)*

**Comment by petermattis** [comment]
I ran a local version of the test. I'll run the actual test before merging.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9703#-:-KTEjTtIXCigz6kkDKn6:b-z0h3et)*

**Comment by cuongdo** [comment]
Also, though the results speak for themselves, I'm still puzzled why Node 1 almost consistently got a monopoly on leases. Maybe it's because node 1 was the leaseholder just before the test data was archived?

**Comment by cuongdo** [comment]
Or is it because node 1 is the only node in the cluster with an empty --join parameter? So before the other nodes join the cluster it's operating as a cluster of 1 node.

**Comment by tschottdorf** [comment]
I pictured that it just got a head start, being the first node and all.

**Comment by petermattis** [comment]
Ditto what @tschottdorf. I bet node 1 starting up first is more likely in the terraform-based test. When I run the test locally, all of the nodes start almost simultaneously.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9703#-:-KTErCetEcZXNe4QnmL6:b2lgm93)*

**Comment by cuongdo** [comment]
yeah, I buy that

**Pull request merged**